packages/xod-project/src/patch.js

```diff
@@ -746,28 +746,57 @@ export const upsertComments = def(
 //
 // =============================================================================
 
+const listLinksNotFromDeferredNodes = def(
+  'listLinksNotFromDeferredNodes :: Patch -> [Link]',
+  R.converge(
+    (deferredNodeIds, links) => R.reject(
+      R.pipe(Link.getLinkOutputNodeId, isAmong(deferredNodeIds)),
+      links
+    ),
+    [
+      R.compose(
+        R.map(Node.getNodeId),
+        R.filter(R.pipe(Node.getNodeType, isDeferNodeType)),
+        listNodes
+      ),
+      listLinks,
+    ]
+  )
+);
+
 /**
  * Returns a topology of nodes in the patch.
+ * defer-* nodes are always placed at the very end.
  *
  * @function getTopology
  * @param {Patch} patch
  * @returns {Either<Error|Array<string>>}
  */
 export const getTopology = def(
   'getTopology :: Patch -> Either Error [NodeId]',
-  R.converge(
-    sortGraph,
-    [
-      R.compose(
-        R.map(Node.getNodeId),
-        listNodes
-      ),
-      R.compose(
-        R.map(Link.getLinkNodeIds),
-        listLinks
-      ),
-    ]
-  )
+  patch => R.compose(
+    R.map(R.compose(
+      R.unnest,
+      R.partition(R.compose(
+        R.complement(isDeferNodeType),
+        Node.getNodeType,
+        R.flip(getNodeByIdUnsafe)(patch)
+      ))
+    )),
+    R.converge(
+      sortGraph,
+      [
+        R.compose(
+          R.map(Node.getNodeId),
+          listNodes
+        ),
+        R.compose(
+          R.map(Link.getLinkNodeIds),
+          listLinksNotFromDeferredNodes
+        ),
+      ]
+    )
+  )(patch)
```

Comment:
The composition now looks quite complex. Perhaps, it is a good idea to split the function into two. Something like:

```
toposortGraph :: Patch -> Either Error [NodeId]
sendDeferNodesToBottom :: Patch -> [NodeId] -> [NodeId]

getTopology = compose(sendDeferNodesToBottom, toposortGraph);
```